main/src/ca/uwaterloo/flix/language/phase/Optimizer.scala

```diff
@@ -0,0 +1,170 @@
+    val constants = root.constants.map {
+      case (name, defn) =>
+        var fn = defn
+
+        // Do simple copy propagation to get rid of targets of direct assignment
+        fn = CopyPropagation.optimize(fn)
+
+        // Replace Unit == Unit checks
+        fn = EliminateUnitChecks.optimize(fn)
+
+        // Clean up by removing code
+        fn = DeadCodeElimination.optimize(fn)
+
+        name -> fn
+    }
+
+    root.copy(constants = constants)
+  }
+
+  object EliminateUnitChecks {
+
+    def optimize(f: Constant): Constant = {
+
+      object ReplaceUnitCheckWithTrue extends ExpressionFolder {
+        override def foldBinary(op: BinaryOperator,
+                                exp1: Expression,
+                                exp2: Expression,
+                                tpe: Type,
+                                loc: SourceLocation): Expression = (op, exp1.tpe) match {
+          // Unit == Unit is always true
+          case (BinaryOperator.Equal, Unit.tpe) => True
+
+          case _ => Binary(op, exp1, exp2, tpe, loc)
+        }
+      }
+
+      f.copy(exp = ReplaceUnitCheckWithTrue.foldExpression(f.exp))
+    }
+
+  }
+
+  object DeadCodeElimination {
+
+    def optimize(f: Constant): Constant = {
+
+      object RemoveKnownConditional extends ExpressionFolder {
+        override def foldIfThenElse(cond: Expression,
+                                    exp1: Expression,
+                                    exp2: Expression,
+                                    tpe: Type,
+                                    loc: SourceLocation): Expression = cond match {
+          // Condition known to be true, so replace whole expression with just the consequent expression
+          case Expression.True => exp1
+
+          // Condition known to be false, so replace whole expression with just the alternative expression
+          case Expression.False => exp2
+
+          // Don't know one way or the other so we don't modify the expression
+          case _ => IfThenElse(cond, exp1, exp2, tpe, loc)
+        }
+      }
+
+      f.copy(exp = RemoveKnownConditional.foldExpression(f.exp))
+    }
+
+  }
+
+  object CopyPropagation {
+
+    def optimize(f: Constant): Constant = {
+
+      /**
+        * Replaces any references to `oldVar` with `newVar`.
+        */
+      class ReplaceVarInExp(oldVar: Name.Ident,
+                            newVar: Name.Ident,
+                            newOffset: scala.Int,
+                            newLoc: SourceLocation) extends ExpressionFolder {
+        override def foldVar(ident: Name.Ident,
+                             offset: scala.Int,
+                             tpe: Type,
+                             loc: SourceLocation): Expression = {
+          if (ident.name == oldVar.name) {
+            Var(newVar, newOffset, tpe, newLoc)
+          } else {
+            Var(ident, offset, tpe, loc)
+          }
+        }
+
+        override def foldMkClosure(lambda: Lambda,
```

Comment:
I'd have to double-check, but I think `MkClosure` nodes have been completely replaced by `MkClosureRef` at this point (i.e. after lambda lifting). So you could replace this with an InternalCompilerError.